src/test/unit/mcmc/chainset_test.cpp

```diff
@@ -0,0 +1,334 @@
+
+  bad.clear();
+  bad.push_back(bernoulli_default);
+  bad.push_back(bernoulli_zeta);
+  EXPECT_THROW(stan::mcmc::chainset fail(bad), std::invalid_argument);
+}
+
+TEST_F(McmcChains, paramNameIndex) {
+  stan::mcmc::chainset chains_csv(eight_schools_1);
+  EXPECT_EQ(1, chains_csv.num_chains());
+  for (int i = 0; i < eight_schools_1.header.size(); i++) {
+    EXPECT_EQ(eight_schools_1.header[i], chains_csv.param_name(i));
+    EXPECT_EQ(i, chains_csv.index(eight_schools_1.header[i]));
+  }
+  EXPECT_THROW(chains_csv.param_name(-5000), std::invalid_argument);
+  EXPECT_THROW(chains_csv.param_name(5000), std::invalid_argument);
+  EXPECT_THROW(chains_csv.index("foo"), std::invalid_argument);
+}
+
+TEST_F(McmcChains, eight_schools_samples) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  Eigen::MatrixXd mu_all = chain_2.samples("mu");
+  EXPECT_EQ(chain_2.num_chains() * chain_2.num_samples(), mu_all.size());
+  mu_all = chain_2.samples(7);
+  EXPECT_EQ(chain_2.num_chains() * chain_2.num_samples(), mu_all.size());
+
+  EXPECT_THROW(chain_2.samples(5000), std::invalid_argument);
+  EXPECT_THROW(chain_2.samples("foo"), std::invalid_argument);
+}
+
+TEST_F(McmcChains, split_rank_normalized_rhat) {
+  stan::mcmc::chainset chain_1(eight_schools_1);
+  EXPECT_EQ(1, chain_1.num_chains());
+
+  // test against R implementation in pkg posterior
+  Eigen::VectorXd rhat_8_schools_1_bulk(10);
+  rhat_8_schools_1_bulk << 1.0012958313, 1.0046136496, 1.0085723580,
+      1.0248629375, 1.0111456620, 1.0004458336, 0.9987162973, 1.0339773469,
+      0.9985612618, 1.0281667351;
+
+  Eigen::VectorXd rhat_8_schools_1_tail(10);
+  rhat_8_schools_1_tail << 1.005676523, 1.009670999, 1.00184184, 1.002222679,
+      1.004148161, 1.003218528, 1.009195353, 1.001426744, 1.003984381,
+      1.025817745;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto rhats = chain_1.split_rank_normalized_rhat(i + 7);
+    EXPECT_NEAR(rhats.first, rhat_8_schools_1_bulk(i), 0.05);
+    EXPECT_NEAR(rhats.second, rhat_8_schools_1_tail(i), 0.05);
+  }
+}
+
+TEST_F(McmcChains, split_rank_normalized_ess) {
+  std::vector<stan::io::stan_csv> eight_schools;
+  eight_schools.push_back(eight_schools_1);
+  eight_schools.push_back(eight_schools_2);
+  stan::mcmc::chainset chain_2(eight_schools);
+  EXPECT_EQ(2, chain_2.num_chains());
+
+  // test against R implementation in pkg posterior (via cmdstanr)
+  Eigen::VectorXd ess_8_schools_bulk(10);
+  ess_8_schools_bulk << 348, 370, 600, 638, 765, 608, 629, 274, 517, 112;
+  Eigen::VectorXd ess_8_schools_tail(10);
+  ess_8_schools_tail << 845, 858, 874, 726, 620, 753, 826, 628, 587, 108;
+
+  for (size_t i = 0; i < 10; ++i) {
+    auto ess = chain_2.split_rank_normalized_ess(i + 7);
+    EXPECT_NEAR(ess.first, ess_8_schools_bulk(i), 5);
+    EXPECT_NEAR(ess.second, ess_8_schools_tail(i), 5);
```

Comment:
Similar question on tolerances -- why don't we expect some exact answer?

Reply:
to compute ESS, the C++ implementation first does rank-normalization, then splits the chains and uses the Boost libraries to compute autocovariance at all lags.
because the underlying libraries may differ, the results differ, but not by much.   anything wildly off would be suspect, but given that the numbers we get for a dataset are relatively close, I think this is OK.